python/mypy
# [needs tests] TypedDict: Recognize declaration of TypedDict('Point', {'x': int, 'y': int}).

The automated tests in this PR still need work but the feature implementation should be good.

## How to test locally

* Add davidfstr/mypy remote

```
$ pwd
/path/to/mypy
$ git remote add davidfstr git@github.com:davidfstr/mypy.git
$ git fetch davidfstr
```

* Checkout tip of this PR

```
$ git checkout declare_typeddict
```

* Create TEST.py

```
$ cat - > TEST.py
from mypy.typing import TypedDict
Point = TypedDict('Point', {'x': int, 'y': int})
point = dict(x=42, y=1337)  # type: Point
^D
```

* Run TEST.py to show it has no runtime errors

```
$ python3 TEST.py
```

* Run mypy on TEST.py

```
$ python3 -m mypy --show-traceback TEST.py 
TEST.py:7: error: Incompatible types in assignment (expression has type Dict[str, int], variable has type "Point")
```

That single error above is expected since this PR only supports declaring the TypedDict type `Point = TypedDict('Point', {'x': int, 'y': int})` but not the TypedDict instance. If you remove the last line (`point = dict(x=42, y=1337)  # type: Point`) then you should get no errors.

## Notes

### Move typing.TypedDict -> mypy.typing.TypedDict?

<s>During the incremental adding of support of TypedDict over several PRs, it would be a lot easier if `TypedDict` was not defined in the official `typing` module but was instead defined in a mypy-specific typing module (like `mypy.typing`). Then after TypedDict was stabilizied, then TypedDict could be added to the official `typing` module. Let me know if you this is a good idea.</s>

Done. It's a lot cleaner now. No more conditional imports.

### More work required on tests

I have tried to implement tests for TypedDict but am getting a failure I do not understand.

```
$ python3 runtests.py mypy.test.testsemanal
PARALLEL 2
SUMMARY  3 tasks selected
passed 2, failed 1, pending 0; running 0                                        

FAILURE  #0 run unit-test mypy.test.testsemanal

main:1: note: In module imported here:
/Users/davidf/Projects/mypy/mypy/typing.py:5: error: Name '__builtins__.dict' is not defined
main:1: note: In module imported here:
/Users/davidf/Projects/mypy/mypy/typing.py:5: error: Name '__builtins__.dict' is not defined
/Users/davidf/Projects/mypy/mypy/typing.py:5: error: INTERNAL ERROR -- please report a bug at https://github.com/python/mypy/issues
/Users/davidf/Projects/mypy/mypy/typing.py:5: note: please use --show-traceback to print a traceback when reporting a bug

SystemExit: 1

test_testsemanal_SemAnalSuite.testCanDefineTypedDictType failed

/Users/davidf/Projects/mypy/mypy/typing.py:5: error: INTERNAL ERROR -- please report a bug at https://github.com/python/mypy/issues
/Users/davidf/Projects/mypy/mypy/typing.py:5: note: please use --show-traceback to print a traceback when reporting a bug

SystemExit: 1

test_testsemanal_SemAnalSuite.testCannotYetAnnotateDictLiteralAsTypedDict failed

2/476 test cases failed.
*** FAILURE ***
```

I am not sure how to pass `--show-traceback` to the test suite, as the error message recommends.

**Edit:** Simplify instructions now that the new `mypy.typing` module to hold `TypedDict` is used in lieu of the standard `typing` module.
